[ConsoleLauncher] Extend stack trace output

olcbean opened the pull request:
Include root cause and suppressed exceptions in the stack trace printed
by the Console Launcher

Resolves #1693.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

sormuras commented on junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java:
```diff
@@ -213,13 +216,46 @@ private void printSource(PrintWriter writer, TestIdentifier testIdentifier) {
+			writer.printf("%s%s%s%n", indentation, caption, throwable);
+		}
+		int duplicates = numberOfCommonFrames(trace, parentTrace);
+		int numLines = (trace.length - duplicates > max) ? max : trace.length - duplicates;
```
Nit: any good name for the difference `trace.length - duplicates`? Could be assigned to local variable used twice here.

sormuras commented on junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java:
```diff
@@ -213,13 +216,46 @@ private void printSource(PrintWriter writer, TestIdentifier testIdentifier) {
+		}
+	}
+
+	private int numberOfCommonFrames(StackTraceElement[] currentTrace, StackTraceElement[] parentTrace) {
```
Interesting computation -- seems to be free of `IndexOutOfBoundsException`s.

sormuras commented on junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java:
```diff
@@ -213,13 +216,46 @@ private void printSource(PrintWriter writer, TestIdentifier testIdentifier) {
+		}
+
+		for (Throwable suppressed : throwable.getSuppressed()) {
+			printStackTrace(writer, trace, suppressed, SUPPRESSED, indentation + TAB, max);
```
Should we track which throwable was already handled to prevent eternal recursions?

sormuras commented on junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java:
```diff
@@ -213,13 +216,46 @@ private void printSource(PrintWriter writer, TestIdentifier testIdentifier) {
+			printStackTrace(writer, trace, suppressed, SUPPRESSED, indentation + TAB, max);
+		}
+		if (throwable.getCause() != null) {
+			printStackTrace(writer, trace, throwable.getCause(), CAUSED_BY, indentation, max);
```
Potential eternal recursion?

olcbean commented:
@sormuras I just rebased and added a note to the release notes. I am not sure about the format / style so any pointers are highly appreciated ;) 
Thanks!

sormuras commented:
Please, rebase again and solve the conflict -- I'll "squash and merge" this PR faster this time.

olcbean commented:
@sormuras rebased and conflicts free ;)

sormuras commented:
Thanks @olcbean!
